tests/shared/core/test_generator.py

```diff
@@ -18,3 +24,59 @@ def test_subsample_array():
 
     assert len(r) == 5
     assert set(r).issubset(t)
+
+
+def test_tracker_with_cached_states_fingerprinting_consistency(domain: Domain):
+    slot = TextSlot(name="name", influence_conversation=True)
+    slot.value = "example"
+    tr1 = rasa.shared.core.generator.TrackerWithCachedStates(
+        "test_sender_id", slots=[slot], domain=domain
+    )
+    tr2 = rasa.shared.core.generator.TrackerWithCachedStates(
+        "test_sender_id", slots=[slot], domain=domain
+    )
+    f1 = tr1.fingerprint()
+    f2 = tr2.fingerprint()
+    assert f1 == f2
+
+
+def test_tracker_with_cached_states_unique_fingerprint(domain: Domain):
+    slot = TextSlot(name="name", influence_conversation=True)
+    slot.value = "example"
+    tr = rasa.shared.core.generator.TrackerWithCachedStates(
+        "test_sender_id", slots=[slot], domain=domain
+    )
+    f1 = tr.fingerprint()
+
+    tracker = DialogueStateTracker("test_sender_id", slots=[slot])
+    event = UserUttered(
+        text="hello",
+        parse_data={
+            "intent": {"id": 2, "name": "greet", "confidence": 0.9604260921478271},
+            "entities": [
+                {"entity": "city", "value": "London"},
+                {"entity": "count", "value": 1},
+            ],
+            "text": "hi",
+            "message_id": "3f4c04602a4947098c574b107d3ccc59",
+            "metadata": {},
+            "intent_ranking": [
+                {"id": 2, "name": "greet", "confidence": 0.9604260921478271},
+                {"id": 1, "name": "goodbye", "confidence": 0.01835782080888748},
+                {"id": 0, "name": "deny", "confidence": 0.011255578137934208},
+            ],
+        },
+    )
+    tracker.update(event, domain=domain)
+    container = deque()
+    container.append(tracker.freeze_current_state(domain.get_active_state(tracker)))
+    tr._states_for_hashing = container
```

Comment:
Not very pleased with accessing a private attribute here, I didn't know how to update the cached states otherwise🤔

Reply:
Does the cached states do anything other than just speed up certain methods? I.e. does it actually hold any unique information not encapsulated by the events and slots etc.?

Reply:
At a first look [here](https://github.com/RasaHQ/rasa/blob/2039cef57780f034da22efcf8c1534d12f3d0ac3/rasa/shared/core/domain.py#L1078) - seems that most of the information here should be captured in events, domain, slots. I can remove this and instead add `self.events` to fingerprintable data?